Source/Particles/PhysicalParticleContainer.cpp

```diff
@@ -2112,3 +2130,129 @@ PhysicalParticleContainer::ContinuousInjection(const RealBox& injection_box)
     const int lev=0;
     AddPlasma(lev, injection_box);
 }
+
+/* \brief Gather fields from FArrayBox exfab, eyfab, ezfab, bxfab, byfab, 
+ * bzfab into arrays of fields on particles Exp, Eyp, Ezp, Bxp, Byp, Bzp.
+ * \param Exp-Bzp: fields on particles.
+ * \param exfab-bzfab: FAB of electric and magnetic fields for particles in pti
+ * \param ngE: number of guard cells for E
+ * \param e_is_nodal: 0 if E is staggered, 1 if E is nodal
+ * \param offset: index of first particle for which fields are gathered
+ * \param np_to_gather: number of particles onto which fields are gathered
+ * \param thread_num: if using OpenMP, thread number
+ * \param lev: level on which particles are located
+ * \param gather_lev: level from which particles gather fields (lev-1) for 
+          particles in buffers.
+ */
+void
+PhysicalParticleContainer::FieldGather(WarpXParIter& pti,
+                                       RealVector& Exp,
+                                       RealVector& Eyp,
+                                       RealVector& Ezp,
+                                       RealVector& Bxp,
+                                       RealVector& Byp,
+                                       RealVector& Bzp,
+                                       FArrayBox const * exfab,
+                                       FArrayBox const * eyfab,
+                                       FArrayBox const * ezfab,
+                                       FArrayBox const * bxfab,
+                                       FArrayBox const * byfab,
+                                       FArrayBox const * bzfab,
+                                       const int ngE, const int e_is_nodal,
+                                       const long offset,
+                                       const long np_to_gather,
+                                       int thread_num,
+                                       int lev,
+                                       int gather_lev)
+{
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE((gather_lev==(lev-1)) ||
+                                     (gather_lev==(lev  )),
+                                     "Gather buffers only work for lev-1");
+    
+    // If no particles, do not do anything
+    if (np_to_gather == 0) return;
+    
+    const std::array<Real,3>& dx = WarpX::CellSize(std::max(gather_lev,0));
+    const Real stagger_shift = e_is_nodal ? 0.0 : 0.5;
```

Comment:
we could even make `e_is_nodal` a template parameter and use `constexpr` to get rid of some registers.
update: oh wait, that's not a kernel.

Reply:
oh, were you asking about this offline earlier? Yes, the transformation I propose to a template param would very likely reduce the register count; we can simply validate this assumption during compile.